### PR TITLE
Создание train_test_split_node #56

### PR DESCRIPTION
**Pull request by DemiusHTTV**
Создал функцию функцию которая автоматически разделяет входящие данные на train и test 
- Добавил в split() валидацию размеров входов (x, y) и параметров train_size/test_size, с последующей  автоподстановкой недостающего размера (или дефолтного теста 25%)
- Добавление перемешки данных, чтобы обучение было хорошим
- Подключил train_test_split_node к Config/Annatations/annotation.py
- Подключил train_test_split_node в Src/Nodes/__init__.py
- Подключил train_test_split_node в Src/Config/node_list.py
- Добавил тесты test_train_test_split.py в Test/ . Тесты были пройдены

### CODE REVIEW

File: Src/Nodes/train_test_split_node.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+from subprocess import Popen, PIPE
+import numpy as np
+
+from Src.Enums import Themes
+from Src.Nodes import DataNode
+
+class TrainTestSplitNode(DataNode):
+    theme_name: Themes = Themes.UTILS
+    @staticmethod
+    def split(x: np.ndarray, y: np.ndarray, test_size: float = 0.25, train_size:float | None = None, random_state: int | None = None, **kwargs):
```
Comment (jezvgg) [inline Src/Nodes/train_test_split_node.py]: Нарушено правило оформления кода:
Между классми \ большими группами кода - 4 пустые строки,
между функциями - 2 пустые строки,
тут их 0

File: Src/Nodes/train_test_split_node.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
+from subprocess import Popen, PIPE
```
Comment (jezvgg) [inline Src/Nodes/train_test_split_node.py]: А эти импорты вообще нахуя если не юзаются?

File: Src/Nodes/train_test_split_node.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            raise ValueError("x должен быть массивом с размерностью не меньше 1")
+        if y is not None and y.shape[0] != n_samples:
+            raise ValueError("Количество образцов в x и y должно быть одинаковым")
+        def _size_to_int(size,n_samples):
+            if isinstance(size,float):
+                return int(size*n_samples)
+            elif isinstance(size,int):
+                return size
+            else:
+                raise ValueError("Размер должен быть числом с плавающей запятой или целым числом")
```
Comment (jezvgg) [inline Src/Nodes/train_test_split_node.py]: ЕЩЁ РАЗ ТЫ БЛ*ТЬ ФУНКЦИЮ ВНУТРИ ФУНКЦИИ ТК ОБЪЯВИШЬ Я ТВОИ РУКИ В ЖОПУ ЗАПИХАЮ

File: Src/Nodes/train_test_split_node.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    theme_name: Themes = Themes.UTILS
+    @staticmethod
+    def split(x: np.ndarray, y: np.ndarray, test_size: float = 0.25, train_size:float | None = None, random_state: int | None = None, **kwargs):
+        x = np.asarray(x) 
+        y = np.asarray(y) if y is not None else None
+        n_samples = x.shape[0]
+        if x.ndim ==0:
+            raise ValueError("x должен быть массивом с размерностью не меньше 1")
+        if y is not None and y.shape[0] != n_samples:
+            raise ValueError("Количество образцов в x и y должно быть одинаковым")
```
Comment (jezvgg) [inline Src/Nodes/train_test_split_node.py]: Вот это молодец, правильно сделал, но проверки нужно расширить на все сценарии. Посмотри как у нас в остальных узлах они сделаны.

File: Src/Nodes/train_test_split_node.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        if test_size is None and train_size is None:
+            test_size = int(0.25 * n_samples)
+            train_size = n_samples - test_size
+        elif test_size is None:
+            test_size = n_samples - train_size
+        elif train_size is None:
+            train_size = n_samples - test_size
+
+        if test_size <= 0 or train_size <= 0 or test_size + train_size > n_samples:
+            raise ValueError("Некорректные размеры тестовой и обучающей выборки")
```
Comment (jezvgg) [inline Src/Nodes/train_test_split_node.py]:
```python
test_size = test_size or n_samples - (train_size or n_samples * 0.75)
train_size = train_size or n_samples - test_size
if tran_size + test_size > n_samples: raise Exception()
```
Код нужно писать максмально простым, чаще всего это достигается путём написания меньшго кода, во тут 3 строчки выполняют тот же функционал, что и твои 9. Чистый код достигаем путём его упрощения.
or - это nullable оператор в нашем случае, если делать None or 3, то значение будет 3.
Продвинутый лайфхак, которым чат гпт пользоваться не умеет.

File: Src/Nodes/train_test_split_node.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+        indices = np.arange(n_samples)
+        rng = np.random.default_rng(random_state)
+        rng.shuffle(indices)
```
Comment (jezvgg) [inline Src/Nodes/train_test_split_node.py]: indices = np.random.shuffle(indices)

вместо 2 строк можно в одну уместить

File: Src/Nodes/train_test_split_node.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
+from subprocess import Popen, PIPE
+import numpy as np
+
+from Src.Enums import Themes
+from Src.Nodes import DataNode
+
+class TrainTestSplitNode(DataNode):
```
Comment (jezvgg) [inline Src/Nodes/train_test_split_node.py]: Тут должна быть ShapeNode, по аналогии с DatasetNode. При чём я бы кстати не реализовал класс, а как раз таки реализовал статическую функция внутри DatasetNode.

File: Src/Nodes/train_test_split_node.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+        indices = np.arange(n_samples)
+        rng = np.random.default_rng(random_state)
+        rng.shuffle(indices)
+        
+        n_train = indices[:train_size]
+        n_test = indices[train_size:train_size+test_size]
+        X_train, X_test = x[n_train], x[n_test]
+        y_train, y_test = y[n_train], y[n_test]
+        return X_train, X_test, y_train, y_test if y is not None else (X_train, X_test)
```
Comment (jezvgg) [inline Src/Nodes/train_test_split_node.py]: Такой тип выходных данных не в каких узлах не поддержвается, везде ожидается np.array

File: Src/Nodes/train_test_split_node.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+            elif len(result) == 4:
+                self.X_train, self.X_test, self.y_train, self.y_test = result
+            else:
+                raise AttributeError("split вернул некорректное количество выходов")
+
+        except Exception as ex:
+            self.raise_error(ex, "Ошибка при split")
+            return False
+
+        return status
```
Comment (jezvgg) [inline Src/Nodes/train_test_split_node.py]: Вообще лишний код, если сделать функцию внутри DatasetNode, то писать не пришлось бы

File: Tests/test_train_test_split.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        np.testing.assert_array_equal(y_test,  [1, 0])
+    def test_mismatched_lengths(self):
+        x = np.array([[1,2],[3,4],[5,9]])
+        y = np.array([0])
+        with self.assertRaises(ValueError) as e:
+            
+            TrainTestSplitNode.split(x,y,test_size=0.25,random_state=42)
+        print(e.exception)
+if __name__ == "__main__":
+    unittest.main()
```
Comment (jezvgg) [inline Tests/test_train_test_split.py]: Нарушены условия оформления кода.
4 строки пусте между классами (или if __name__ == '__main__' и классом)
2 строки пустые межу функциями
Кстати if __name__ == '__main__' вместе с вызовом еста тут не нужен, можно без него обойтись

File: Src/Config/node_list.py [hunk 11]
```diff
@@ -76,6 +76,18 @@
                 annotations = {},
                 input=Single[DataNode]
             ),
+            NodeAnnotation(
+                label="Train test split",   
+                node_type= TrainTestSplitNode,
+                logic = TrainTestSplitNode.split,
+                annotations = {
+                        "test_size": Parameter(AttrType.INPUT, AFloat, default=0.25),
+                        "train_size": Parameter(AttrType.INPUT, AFloat | AInteger, default=None)
```
Comment (jezvgg) [inline Src/Config/node_list.py]: TrainTestSplitNode нету в nodes.__init__, так чо он сюда не импортируется и не работает.
Кроме ого, AFloat и AInteger, это наши собственные аннотации, они не просто указывают тип, но ещё и гоорят о том, как создавать узлы, а когда ты делаешь |, то неявно создаёшь тип аннотации Union, который не знает как из себя создавать узлы в констркторе, из-за чего всё тоже ломается.

**Review by jezvgg — CHANGES_REQUESTED** [review]
1. Оформи код, как договаривались.
2. Нужно 2 метода - один делает train_test_split, а второй разделяет один общий массив на X и y.
3. train_test_split реализуй в DatasetNode и по аналогии с функцией что там уже есть
4. Также узел должен получить на вход X и y, а у тебя он сейчас их не получает или получает через Input, что не правильно.
5. Постарайся сократить код до 10 строчек внутри функции примерно, потому что чем меньше кода - тем он проще.

**Comment by DemiusHTTV** [comment]
Все правки исправил, но появилась проблема. Функции по отдельности "train_test_split" и "splitXy" работают исправно, но при сборке проекта появляется эта ошибка, не знаю как исправить
image

Если коротко как я понял - узел "Split X/y" возвращает кортеж (X, y), но DataNode.compile не раскладывает его в атрибуты X и y. ANode при получении связей ищет поля у узла, а их нет.

Единственное что я могу придумать это создать свой compile, но это значит узел придется создавать отдельный SplitXY -какой нибудь, но я же работаю только в datasets_node. Ну либо в Datasets compile переписать но это опасно там точно все полетит

File: Src/Config/node_list.py [hunk 12]
```diff
@@ -76,6 +76,32 @@
...
             ),
+            NodeAnnotation(
+            label="Train test split",
+            node_type=DataNode,                    
+            logic=DatasetNode.train_test_split,     
+            annotations={
+            "x": Parameter(AttrType.INPUT, ANode[Single[DataNode]]),
+            "y": Parameter(AttrType.INPUT, ANode[Single[DataNode]]),
+            "test_size": Parameter(AttrType.INPUT, AFloat, default=0.25),
+            "train_size": Parameter(AttrType.INPUT, AFloat, default=None),
```
Comment (jezvgg) [inline Src/Config/node_list.py]: Тебе не нужен train_size, одного test_size хватит. Так и функция станет проще.

File: Src/Config/node_list.py [hunk 13]
```diff
@@ -76,6 +76,32 @@
...
+            output=DataNode,  
+            ),
+            NodeAnnotation(
+                label="Split X/y",
+                node_type=DataNode,           
+                logic=DatasetNode.split_Xy,
+                annotations={
+                "data": Parameter(AttrType.INPUT, ANode[DataNode]),
+                 "X": Parameter(AttrType.OUTPUT, ANode[DataNode]),
+                 "y": Parameter(AttrType.OUTPUT, ANode[DataNode]),
```
Comment (jezvgg) [inline Src/Config/node_list.py]: А где параметр выбора y? y - один из столбцов, которые мы должны указать

File: Src/Nodes/dataset_node.py [hunk 14]
```diff
@@ -49,6 +49,74 @@ def open_data(dataset:str) -> Dataset:
...
+
+
+    @staticmethod
+    def train_test_split(x: np.ndarray, y: np.ndarray, test_size: float = 0.25, train_size:float | None = None, random_state: int | None = None, **kwargs):
+        '''
+
+        Разделяет полученные данные на train и на test и перемешивает их для лучшего обучения
+
+        '''
+        x, y =np.asarray(x) , np.asarray(y) if y is not None else None
```
Comment (jezvgg) [inline Src/Nodes/dataset_node.py]: У тебя x и y обязательные аргументы, оба из которых точно будут np.array, лишнии проверки

File: Src/Nodes/dataset_node.py [hunk 15]
```diff
@@ -49,6 +49,74 @@ def open_data(dataset:str) -> Dataset:
...
+        n_samples = x.shape[0]
+        if x.ndim ==0:
+            raise ValueError("x должен быть массивом с размерностью не меньше 1")
+        if y is not None and y.shape[0] != n_samples:
+            raise ValueError("Количество образцов в x и y должно быть одинаковым")
+        
+        test_size = DatasetNode._size_to_int(test_size, n_samples) if test_size is not None else n_samples - train_size
+        train_size = DatasetNode._size_to_int(train_size, n_samples) if train_size is not None else n_samples - test_size
```
Comment (jezvgg) [inline Src/Nodes/dataset_node.py]: У тебя они оба в node_list ввиде AFloat, зачем ты их повторно приводишь к int?

File: Src/Nodes/dataset_node.py [hunk 16]
```diff
@@ -49,6 +49,74 @@ def open_data(dataset:str) -> Dataset:
...
+
+        test_size = test_size or n_samples - (test_size or train_size - n_samples * 0.75)
+        train_size = train_size or n_samples - test_size
```
Comment (jezvgg) [inline Src/Nodes/dataset_node.py]: Можно вообще эти рассчёты убрать, если убрать train_size. train_size нужен просто для удобства, но это удобство работает только с программировании, а у нас визуальный конструктор и дополнительные параметры только усложняют понимание.

File: Src/Nodes/dataset_node.py [hunk 17]
```diff
@@ -49,6 +49,74 @@ def open_data(dataset:str) -> Dataset:
...
+        if train_size + test_size > n_samples: raise Exception("Некорректные размеры тестовой и обучающей выборки")
+
+        indices = np.arange(n_samples)
+        rng = np.random.default_rng(random_state)
+        rng.shuffle(indices)
```
Comment (jezvgg) [inline Src/Nodes/dataset_node.py]: В прошлый раз указывал, что можно в 2 строчки написать

File: Src/Nodes/dataset_node.py [hunk 18]
```diff
@@ -49,6 +49,74 @@ def open_data(dataset:str) -> Dataset:
...
+        
+        n_train = indices[:train_size]
+        n_test = indices[train_size:train_size+test_size]
+        X_train, X_test = x[n_train], x[n_test]
+        y_train, y_test = y[n_train], y[n_test]
+        return X_train, X_test, y_train, y_test if y is not None else (X_train, X_test)
```
Comment (jezvgg) [inline Src/Nodes/dataset_node.py]: Логично, что если у основной функции возращается объект типа Dataset, то и тут должно быть также.

File: Src/Nodes/dataset_node.py [hunk 19]
```diff
@@ -49,6 +49,74 @@ def open_data(dataset:str) -> Dataset:
...
         return Dataset(X_train, y_train, X_test, y_test, X_train.shape)
+    
+    @staticmethod
+    def split_Xy(data: np.ndarray):
+        '''
+            Разделение данных на X и y
+
+        '''
+
+        data = np.asarray(data) if data is not None else Exception()
```
Comment (jezvgg) [inline Src/Nodes/dataset_node.py]: else Exception() что делает по твоему? Чтоб вызвать ошибку нужен raise. При этом у ошибки даже нет описания.

File: Src/Nodes/dataset_node.py [hunk 20]
```diff
@@ -49,6 +49,74 @@ def open_data(dataset:str) -> Dataset:
...
+        data = np.squeeze(data, axis=0) if data.ndim > 2 and data.shape[0] == 1 else np.asarray(data)
+        if data.ndim != 2 or data.shape[1] < 2:
+             raise ValueError(f"Нужно 2D и ≥2 колонки, пришло {data.shape}")
+
+        X = data[:, :-1]
+        y = data[:, -1]
```
Comment (jezvgg) [inline Src/Nodes/dataset_node.py]: Ну вот, у - не всегда последняя колонка, нужно указывать номер этой колонки

File: Src/Nodes/train_test_split_node.py
Comment (jezvgg) [inline Src/Nodes/train_test_split_node.py:1]: Так а этот файл нужен удалить, зачем он остался, не используется же

**Review by jezvgg — CHANGES_REQUESTED** [review]
Не работает потому что неправильный выход у функций твоих. Должен быть заполненый объект типа Dataset, а также выходные названия атрибутов узлов, должны совпадать с названиями полей в классе узла.
Просто посиди, разберись как эти узлы устроенны. 
В следующий раз PR отправляй только когда запуститься оно стабильно и хорошо.

File: Src/Nodes/utils_node.py [hunk 21]
```diff
@@ -1,23 +1,34 @@
...
+
+    @staticmethod
+    def save_txt(data, fname="result.txt", fmt="%g"):
+        """
+        Обёртка над np.savetxt с более читаемым форматированием чисел
+        """
+        try:
+            np.savetxt(fname, data, fmt=fmt)
+        except Exception as ex:
+            raise Exception(f"Непредвиденная ошибка с записью в файл: {ex}")
```
Comment (jezvgg) [inline Src/Nodes/utils_node.py]: Функция лишняя, можно просто в node_list сделать lambda data, fname: np.save_txt(data, fname, fmt="%g")
